Assumptions, the environment for running this notebook was arrived at by building a dedicated docker file.

https://cloud.docker.com/repository/registry-1.docker.io/russelljarvis/nuo

You can run use dockerhub to get the appropriate file, and launch this notebook using Kitematic.

# Import libraries
To keep the standard running version of minimal and memory efficient, not all available packages are loaded by default. In the cell below I import a mixture common python modules, and custom developed modules associated with NeuronUnit (NU) development

In [1]:

import numpy as np
import os
import pickle
import pandas as pd
from neuronunit.tests.fi import RheobaseTestP
from neuronunit.optimization.model_parameters import reduced_dict, reduced_cells  
from neuronunit.optimization import optimization_management as om
from sciunit import scores# score_type 

from neuronunit.optimization.data_transport_container import DataTC
from neuronunit.tests.fi import RheobaseTestP# as discovery
from neuronunit.optimization.optimization_management import dtc_to_rheo, format_test, nunit_evaluation
import quantities as pq
from neuronunit.models.reduced import ReducedModel
from neuronunit.optimization.model_parameters import model_params, path_params
LEMS_MODEL_PATH = path_params['model_path']
list_to_frame = []
from neuronunit.tests.fi import RheobaseTestP

# The Izhiketich model is instanced using some well researched parameter sets.


First lets get the points in parameter space, that Izhikich himself has published about. These points are used by other scientists and the [Open Source Brain project](http://opensourcebrain.org/projects/izhikevichmodel?explorer=https%3A%2F%2Fraw.githubusercontent.com%2FOpenSourceBrain%2FIzhikevichModel%2Fsample%2FosbSessions%2Fsample%2FSample_Session.json)  to establish between model reproducibility. The itial motivating factor for choosing these points as constellations, of all possible parameter space subsets, is that these points where initially tuned and used as best guesses for matching real observed experimental recordings.

In [2]:
explore_param = {k:(np.min(v),np.max(v)) for k,v in reduced_dict.items()}

## Get the experimental Data pertaining to four different classes or neurons, that can constrain models.
Next we get some electro physiology data for four different classes of cells that are very common targets of scientific neuronal modelling. We are interested in finding out what are the most minimal, and detail reduced, low complexity model equations, that are able to satisfy 

Below are some of the data set ID's I used to query neuroelectro.
To save time for the reader, I prepared some data earlier to save time, and saved the data as a pickle, pythons preferred serialisation format.

The interested reader can find some methods for getting cell specific ephys data from neuroelectro in a code file (neuronunit/optimization/get_neab.py) 


In [3]:
purkinje ={"id": 18, "name": "Cerebellum Purkinje cell", "neuron_db_id": 271, "nlex_id": "sao471801888"}
fi_basket = {"id": 65, "name": "Dentate gyrus basket cell", "neuron_db_id": None, "nlex_id": "nlx_cell_100201"}
pvis_cortex = {"id": 111, "name": "Neocortex pyramidal cell layer 5-6", "neuron_db_id": 265, "nlex_id": "nifext_50"}
#This olfactory mitral cell does not have datum about rheobase, current injection values.
olf_mitral = {"id": 129, "name": "Olfactory bulb (main) mitral cell", "neuron_db_id": 267, "nlex_id": "nlx_anat_100201"}
ca1_pyr = {"id": 85, "name": "Hippocampus CA1 pyramidal cell", "neuron_db_id": 258, "nlex_id": "sao830368389"}
pipe = [ fi_basket, ca1_pyr, purkinje,  pvis_cortex]

In [4]:
electro_tests = []
obs_frame = {}
test_frame = {}

from neuronunit.optimization import get_neab
try: 

    electro_path = str(os.getcwd())+'all_tests.p'

    assert os.path.isfile(electro_path) == True
    with open(electro_path,'rb') as f:
        (obs_frame,test_frame) = pickle.load(f)

except:
    for p in pipe:
        p_tests, p_observations = get_neab.get_neuron_criteria(p)
        obs_frame[p["name"]] = p_observations#, p_tests))
        test_frame[p["name"]] = p_tests#, p_tests))
    electro_path = str(os.getcwd())+'all_tests.p'
    with open(electro_path,'wb') as f:
        pickle.dump((obs_frame,test_frame),f)

# Cast the tabulatable data to pandas data frame
There are many among us who prefer potentially tabulatable data to be encoded in pandas data frame.

In [5]:
for k,v in test_frame.items():
    if "olf_mit" not in k:
        obs = obs_frame[k]
        v[0] = RheobaseTestP(obs['Rheobase'])
df = pd.DataFrame.from_dict(obs_frame)
print(test_frame.keys())

dict_keys(['Dentate gyrus basket cell', 'Hippocampus CA1 pyramidal cell', 'Neocortex pyramidal cell layer 5-6', 'Cerebellum Purkinje cell'])


In the data frame below, you can see many different cell types

In [6]:
df


Cerebellum Purkinje cell  \
Cell Capacitance            {'mean': 620.2725 pF, 'n': 4, 'std': 261.30303...   
Input Resistance            {'mean': 142.057692307692 Mohm, 'n': 13, 'std'...   
Membrane Time Constant                                                    NaN   
Resting membrane potential  {'mean': -61.5916666666667 mV, 'n': 6, 'std': ...   
Rheobase                    {'mean': 680.794444444444 pA, 'n': 3, 'std': 4...   
Spike Amplitude             {'mean': 71.2308333333333 mV, 'n': 6, 'std': 9...   
Spike Half-Width            {'mean': 0.41412962962963 ms, 'n': 9, 'std': 0...   
Spike Threshold             {'mean': -46.8947619047619 mV, 'n': 7, 'std': ...   

                                                    Dentate gyrus basket cell  \
Cell Capacitance                                                          NaN   
Input Resistance            {'mean': 222.475 Mohm, 'n': 4, 'std': 168.2911...   
Membrane Time Constant             {'mean': 13.25 ms, 'n': 2, 'std': 3.25 ms}   
Resting membrane potential  {'mean': -62.3 mV, 'n': 3, 'std': 4.0472212689...   
Rheobase                            {'mean': 106.7 pA, 'n': 1, 'std': 0.0 pA}   
Spike Amplitude                      {'mean': 78.6 mV, 'n': 2, 'std': 5.7 mV}   
Spike Half-Width            {'mean': 0.9975 ms, 'n': 4, 'std': 0.442175021...   
Spike Threshold                     {'mean': -38.4 mV, 'n': 2, 'std': 3.4 mV}   

                                               Hippocampus CA1 pyramidal cell  \
Cell Capacitance            {'mean': 89.7960714285714 pF, 'n': 14, 'std': ...   
Input Resistance            {'mean': 107.080327644332 Mohm, 'n': 113, 'std...   
Membrane Time Constant      {'mean': 24.5021946169772 ms, 'n': 46, 'std': ...   
Resting membrane potential  {'mean': -65.2261863636364 mV, 'n': 110, 'std'...   
Rheobase                    {'mean': 189.24 pA, 'n': 17, 'std': 287.163664...   
Spike Amplitude             {'mean': 86.364525297619 mV, 'n': 64, 'std': 1...   
Spike Half-Width            {'mean': 1.31895278450363 ms, 'n': 59, 'std': ...   
Spike Threshold             {'mean': -47.5985714285714 mV, 'n': 70, 'std':...   

                                           Neocortex pyramidal cell layer 5-6  
Cell Capacitance            {'mean': 150.584166666667 pF, 'n': 12, 'std': ...  
Input Resistance            {'mean': 120.672073643411 Mohm, 'n': 86, 'std'...  
Membrane Time Constant      {'mean': 15.7342424242424 ms, 'n': 33, 'std': ...  
Resting membrane potential  {'mean': -68.2481434599156 mV, 'n': 79, 'std':...  
Rheobase                    {'mean': 213.849583333333 pA, 'n': 32, 'std': ...  
Spike Amplitude             {'mean': 80.4351020408164 mV, 'n': 49, 'std': ...  
Spike Half-Width            {'mean': 1.20769387755102 ms, 'n': 49, 'std': ...  
Spike Threshold             {'mean': -42.7357232704403 mV, 'n': 53, 'std':...

In [7]:
df['Hippocampus CA1 pyramidal cell']

# enable R style caching.
#!pip install git+https://github.com/rossant/ipycache
#%load_ext ipycache

Cell Capacitance              {'mean': 89.7960714285714 pF, 'n': 14, 'std': ...
Input Resistance              {'mean': 107.080327644332 Mohm, 'n': 113, 'std...
Membrane Time Constant        {'mean': 24.5021946169772 ms, 'n': 46, 'std': ...
Resting membrane potential    {'mean': -65.2261863636364 mV, 'n': 110, 'std'...
Rheobase                      {'mean': 189.24 pA, 'n': 17, 'std': 287.163664...
Spike Amplitude               {'mean': 86.364525297619 mV, 'n': 64, 'std': 1...
Spike Half-Width              {'mean': 1.31895278450363 ms, 'n': 59, 'std': ...
Spike Threshold               {'mean': -47.5985714285714 mV, 'n': 70, 'std':...
Name: Hippocampus CA1 pyramidal cell, dtype: object

# Tweak Izhikitich equations
with educated guesses based on information that is already encoded in the predefined experimental observations.

In otherwords use information that is readily amenable into hardcoding into equations 

Select out the 'Neocortex pyramidal cell layer 5-6' below, as a target for optimization

In [8]:
free_params = ['a','b','k','c','C','d','vPeak','vr']
hc_ = reduced_cells['RS']
hc_['vr'] = -65.2261863636364
hc_['vPeak'] = hc_['vr'] + 86.364525297619
explore_param['C'] = (hc_['C']-20,hc_['C']+20)
explore_param['vr'] = (hc_['vr']-5,hc_['vr']+5)
use_test = test_frame["Neocortex pyramidal cell layer 5-6"]

#for t in use_test[::-1]:
#    t.score_type = scores.RatioScore
test_opt = {}

with open('data_dump.p','wb') as f:
    pickle.dump(test_opt,f)

In [9]:

use_test[0].observation
print(use_test[0].name)

rtp = RheobaseTestP(use_test[0].observation)
use_test[0] = rtp
print(use_test[0].observation)

RheobaseTestP
{'mean': array(213.849583333333) * pA, 'n': 32, 'std': array(170.452454715608) * pA}


The code in the next two cells below is used to create a sparse grid sampling over the parameter space, using the published and well corrobarated parameter points, from Izhikitch publications, and the Open Source brain

In [10]:
reduced_cells.keys()
test_frame.keys()


dict_keys(['Dentate gyrus basket cell', 'Hippocampus CA1 pyramidal cell', 'Neocortex pyramidal cell layer 5-6', 'Cerebellum Purkinje cell'])

In [12]:
df = pd.DataFrame(index=list(test_frame.keys()),columns=list(reduced_cells.keys()))

try:
    assert 1==2
    with open('Izh_seeds.p','rb') as f:
        seeds = pickle.load(f)
    #assert seeds is not None

except:
    # print('exceptional circumstances pickle file does not exist, rebuilding sparse grid for Izhikich')
    # Below we perform a sparse grid sampling over the parameter space, using the published and well corrobarated parameter points, from Izhikitch publications, and the Open Source brain, shows that without optimization, using off the shelf parameter sets to fit real-life biological cell data, does not work so well.

    for k,v in reduced_cells.items():
        temp = {}
        temp[str(v)] = {}
        dtc = DataTC()
        dtc.tests = use_test
        dtc.attrs = v
        dtc.backend = 'RAW'
        dtc.cell_name = 'vanilla'
        for key, use_test in test_frame.items():
            dtc.tests = use_test
            dtc = dtc_to_rheo(dtc)
            dtc = format_test(dtc)
            if dtc.rheobase is not None:
                if dtc.rheobase!=-1.0:
                    dtc = nunit_evaluation(dtc)
            df[k][str(key)] = dtc.get_ss()

    best_params = {}
    for index, row in df.iterrows():
        best_params[index] = row == row.min()
        best_params[index] = best_params[index].to_dict()


    seeds = {}
    for k,v in best_params.items():
        for nested_key,nested_val in v.items():
            if True == nested_val:
                seed = reduced_cells[nested_key]
                seeds[k] = seed
    with open('Izh_seeds.p','wb') as f:
        pickle.dump(seeds,f)




exceptional circumstances pickle file does not exist, rebuilding sparse grid for Izhikich


/opt/conda/lib/python3.5/site-packages/sciunit/scores/complete.py:68: RuntimeWarning: divide by zero encountered in true_divide
  value = (p_value - o_mean)/o_std
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c


As I said above: 'A sparse grid sampling over the parameter space, using the published and well corrobarated parameter points, from Izhikitch publications, and the Open Source brain', The output from the grid shows that without optimization, using off the shelf parameter sets to fit real-life biological cell data, does not work so well.

In [13]:
df

RS       IB      LTS       TC  \
Dentate gyrus basket cell           5.33575  6.44702  6.03769  6.27535   
Hippocampus CA1 pyramidal cell       4.2652  5.84565  5.70527  5.74633   
Neocortex pyramidal cell layer 5-6  4.66967  5.83884  6.21074  6.36435   
Cerebellum Purkinje cell            6.04342  5.95173  5.47237  5.69825   

                                   TC_burst  
Dentate gyrus basket cell           6.38901  
Hippocampus CA1 pyramidal cell      6.06554  
Neocortex pyramidal cell layer 5-6   6.5461  
Cerebellum Purkinje cell            5.53656

As can be seen none of these well researched parameter sets correspond to good fits, with experimental data about waveform shape. These bad fit's warrant the use of an optimizer, which can efficiently, sparsely, and intelligently explore the parameter space for better model fits. None of the models score higher than:

```4.2652```, and the best is ```5.47237``` in the specific case of the cerebellum purkinje cell. 
and None of the first guess parameter set were very good for the cerebellum Purkinje cell. It should be easy to find a model parameterization that scores better than ```5.47```. A first guess. 

The best NeuronUnit score values found via this sparse grid sampling, are used to seed the first generation of the genetic algorithm, as one of the great motivators of GA optimization, is not to waste any computation time.


We now optimize using it to show what can be achieved with the GA.

In [22]:


try:
    with open('multi_objective_izhi.p','rb') as f:
        test_opt = pickle.load(f)
except:
    MU = 6
    NGEN = 150


    seed = seeds[key]
    print(seed)
    ga_out, _ = om.run_ga(explore_param,NGEN,use_test,free_params=free_params, NSGA = True, MU = MU,seed_pop = seed, model_type = str('RAW'))

    test_opt =  {str('multi_objective_izhi')+str('_ga_out'):ga_out}
    with open('multi_objective_izhi.p','wb') as f:
        pickle.dump(test_opt,f)




In [37]:
for k,v in test_opt.items():
    best_model = v['pf'][0]
summed = best_model.dtc.get_ss()
print('its {2} that {0} is significantly less than {1}'.format(summed,5.47,summed<5.47))
print(key)

its True that 3.1302660755008325 is significantly less than 5.47
Cerebellum Purkinje cell


In [42]:
key = str('Neocortex pyramidal cell layer 5-6')


try:
    with open('multi_objective_izhi'+str(key)+'.p','rb') as f:
        test_opt = pickle.load(f)
except:
    MU = 6
    NGEN = 150
    seed = seeds[key]
    use_test = test_frame[key]
    print(seed)
    ga_out, _ = om.run_ga(explore_param,NGEN,use_test,free_params=free_params, NSGA = True, MU = MU,seed_pop = seed, model_type = str('RAW'))

    test_opt =  {str('multi_objective_izhi')+str(key):ga_out}
    with open('multi_objective_izhi'+str(key)+'.p','wb') as f:
        pickle.dump(test_opt,f)



{'vt': -40, 'c': -50, 'vr': -65.2261863636364, 'vPeak': 21.138338933982595, 'C': 100, 'd': 100, 'k': 0.7, 'a': 0.03, 'b': -2}


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:gen	nevals	avg    	std     	min    	max    
1  	7     	5.22203	0.821745	3.72429	6.03463
2  	5     	5.21212	0.735291	3.97883	6.03463


true minimum 3.72428861926


INFO:__main__:3  	4     	5.13677	0.626946	4.25394	5.65035


true minimum 3.72428861926


INFO:__main__:4  	5     	5.19984	0.430726	4.2626 	5.49532


true minimum 3.72428861926


INFO:__main__:5  	5     	5.44865	0.146973	5.19086	5.64343


true minimum 3.72428861926


INFO:__main__:6  	5     	5.16537	0.625444	3.86973	5.67245


true minimum 3.72428861926


/opt/conda/lib/python3.5/site-packages/deap/tools/crossover.py:324: RuntimeWarning: invalid value encountered in double_scalars
  beta_q = (1.0 / (2.0 - rand * alpha))**(1.0 / (eta + 1))
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:7  	6     	5.47294	0.286328	4.88053	5.71681


true minimum 3.72428861926


INFO:__main__:8  	5     	5.1169 	0.84742 	3.22897	5.61862


true minimum 3.72428861926


INFO:__main__:9  	5     	4.74357	0.795977	3.72429	5.57846


true minimum 3.72428861926


INFO:__main__:10 	4     	4.66108	0.967007	3.24729	5.67859


true minimum 3.24728879435


INFO:__main__:11 	5     	5.43769	0.347115	4.67548	5.69242


true minimum 3.24728879435


INFO:__main__:12 	2     	4.99029	0.403366	4.6111 	5.54829


true minimum 3.24728879435


INFO:__main__:13 	5     	4.78651	0.605765	3.59511	5.54829


true minimum 3.24728879435


INFO:__main__:14 	6     	4.66135	0.838683	3.3331 	5.74215


true minimum 3.24728879435


INFO:__main__:15 	6     	4.84525	0.473134	4.21751	5.55403


true minimum 3.24728879435


INFO:__main__:16 	6     	4.14798	0.961541	3.32742	5.55374


true minimum 3.24728879435


INFO:__main__:17 	5     	4.46228	0.904249	3.32742	5.37721


true minimum 3.24728879435


INFO:__main__:18 	4     	3.9865 	0.959449	3.09455	5.34938


true minimum 3.24728879435


INFO:__main__:19 	6     	4.36213	0.923614	3.30176	5.30928


true minimum 3.24728879435


INFO:__main__:20 	6     	4.89491	0.710129	3.32925	5.37684


true minimum 3.24728879435


INFO:__main__:21 	3     	4.96855	0.440792	4.00273	5.27258


true minimum 3.24728879435


INFO:__main__:22 	5     	3.99606	0.777385	3.24729	5.14799


true minimum 3.24728879435


INFO:__main__:23 	6     	4.10263	0.943001	2.99182	5.37337


true minimum 2.99182182811


INFO:__main__:24 	5     	4.62393	0.589267	3.55272	5.20436


true minimum 2.99182182811


INFO:__main__:25 	5     	4.146  	0.595203	3.24729	5.22062


true minimum 2.99182182811


INFO:__main__:26 	5     	4.30695	0.785958	3.30176	5.22537


true minimum 2.99182182811


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:27 	6     	4.31459	1.00617 	2.99182	5.56744


true minimum 2.99182182811


INFO:__main__:28 	6     	4.40688	0.562529	3.32854	5.14262


true minimum 2.99182182811


INFO:__main__:29 	5     	4.542  	0.619648	3.27655	5.26655


true minimum 2.99182182811


INFO:__main__:30 	6     	4.53346	0.334873	4.17312	5.1105 


true minimum 2.99182182811


INFO:__main__:31 	4     	4.43568	0.303636	4.17312	5.03305


true minimum 2.99182182811


INFO:__main__:32 	6     	4.61656	0.35095 	4.13065	5.16013


true minimum 2.99182182811


INFO:__main__:33 	5     	4.29056	0.633806	2.99182	4.99444


true minimum 2.99182182811


INFO:__main__:34 	6     	4.41131	0.552691	3.3616 	5.08154


true minimum 2.99182182811


INFO:__main__:35 	5     	4.12985	0.659809	3.18095	4.77331


true minimum 2.99182182811


INFO:__main__:36 	6     	4.52503	0.343159	3.86635	4.93116


true minimum 2.99182182811


INFO:__main__:37 	6     	4.49488	0.434338	3.56517	4.93141


true minimum 2.99182182811


INFO:__main__:38 	6     	4.30537	0.559633	3.36456	5.04942


true minimum 2.99182182811


INFO:__main__:39 	4     	4.32108	0.422403	3.81848	4.88364


true minimum 2.99182182811


INFO:__main__:40 	6     	4.2473 	0.523654	3.38556	4.81716


true minimum 2.99182182811


INFO:__main__:41 	4     	4.52398	0.556695	3.38556	5.12355


true minimum 2.99182182811


INFO:__main__:42 	6     	4.7039 	0.401336	3.85379	5.06121


true minimum 2.99182182811


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:43 	5     	4.36727	0.436229	3.67472	4.79182


true minimum 2.99182182811


INFO:__main__:44 	5     	3.91217	0.521324	3.39523	4.67254


true minimum 2.99182182811


INFO:__main__:45 	5     	4.50612	0.6599  	3.37202	5.22707


true minimum 2.99182182811


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:46 	6     	4.46049	0.704344	3.40209	5.25458


true minimum 2.99182182811


INFO:__main__:47 	6     	4.09745	0.841532	2.99182	5.4212 


true minimum 2.99182182811


INFO:__main__:48 	5     	4.78211	0.744664	3.2739 	5.55851


true minimum 2.99182182811


INFO:__main__:49 	6     	4.66692	0.674102	3.79381	5.62961


true minimum 2.99182182811


INFO:__main__:50 	5     	4.6925 	0.229621	4.44175	5.15158


true minimum 2.99182182811


INFO:__main__:51 	6     	4.87492	0.136585	4.65136	5.03999


true minimum 2.99182182811


INFO:__main__:52 	6     	4.62645	0.279176	4.13133	4.96709


true minimum 2.99182182811


INFO:__main__:53 	6     	4.71274	0.290041	4.4209 	5.28303


true minimum 2.99182182811


INFO:__main__:54 	4     	4.03018	0.749879	2.99182	4.73957


true minimum 2.99182182811


INFO:__main__:55 	4     	4.57711	0.507153	3.59472	5.07764


true minimum 2.99182182811


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:56 	5     	4.965  	0.353734	4.37404	5.44207


true minimum 2.99182182811


INFO:__main__:57 	4     	4.94033	0.290541	4.49424	5.33279


true minimum 2.99182182811


INFO:__main__:58 	4     	4.01939	0.874808	2.99182	4.95734


true minimum 2.99182182811


INFO:__main__:59 	4     	4.58366	0.762433	3.35157	5.40481


true minimum 2.99182182811


INFO:__main__:60 	6     	3.8724 	0.706549	3.15745	5.30354


true minimum 2.99182182811


INFO:__main__:61 	5     	4.04575	0.794679	3.14408	5.30354


true minimum 2.99182182811


INFO:__main__:62 	6     	4.09405	0.896957	3.14351	5.7101 


true minimum 2.99182182811


INFO:__main__:63 	6     	4.297  	0.856857	2.96964	5.21997


true minimum 2.96964367379


INFO:__main__:64 	6     	4.30922	0.66727 	3.29948	5.10631


true minimum 2.96964367379


INFO:__main__:65 	5     	3.78477	0.682947	3.31091	5.09401


true minimum 2.96964367379


INFO:__main__:66 	6     	4.46096	0.522407	3.89602	5.32886


true minimum 2.96964367379


INFO:__main__:67 	5     	4.49385	0.428289	3.91133	5.20338


true minimum 2.96964367379


INFO:__main__:68 	6     	4.39372	0.498931	3.91133	5.15897


true minimum 2.96964367379


INFO:__main__:69 	6     	4.58499	0.393384	3.98095	5.2565 


true minimum 2.96964367379


INFO:__main__:70 	6     	3.96903	0.443902	3.24159	4.52028


true minimum 2.96964367379


INFO:__main__:71 	5     	4.24541	0.390298	3.96593	5.07562


true minimum 2.96964367379


INFO:__main__:72 	5     	3.5692 	0.638318	2.81818	4.60251


true minimum 2.81817662327


INFO:__main__:73 	6     	3.91753	0.512632	3.24758	4.94453


true minimum 2.81817662327


INFO:__main__:74 	4     	3.62826	0.985904	2.81818	5.48352


true minimum 2.81817662327


INFO:__main__:75 	6     	3.85956	0.707637	2.88205	4.66586


true minimum 2.81817662327


INFO:__main__:76 	6     	3.6798 	0.631453	2.81818	4.27577


true minimum 2.81817662327


INFO:__main__:77 	6     	4.15053	0.165328	3.85328	4.38312


true minimum 2.81817662327


INFO:__main__:78 	6     	3.4361 	0.53812 	2.76358	4.20687


true minimum 2.76357717108


INFO:__main__:79 	5     	3.69135	0.331072	3.35456	4.30059


true minimum 2.76357717108


INFO:__main__:80 	6     	4.05232	0.710832	2.76358	5.10213


true minimum 2.76357717108


INFO:__main__:81 	5     	4.15533	0.726237	2.75113	4.96021


true minimum 2.76357717108


INFO:__main__:82 	6     	3.58236	0.452427	2.8718 	4.37413


true minimum 2.76357717108


INFO:__main__:83 	6     	3.60259	0.589934	2.81867	4.64001


true minimum 2.76357717108


INFO:__main__:84 	6     	3.63368	0.613307	3.01626	4.76451


true minimum 2.76357717108


INFO:__main__:85 	6     	4.14238	0.754201	2.82455	4.8034 


true minimum 2.76357717108


INFO:__main__:86 	4     	3.92542	0.729978	3.02329	4.98257


true minimum 2.76357717108


INFO:__main__:87 	5     	3.51059	0.663082	2.94123	4.82076


true minimum 2.76357717108


INFO:__main__:88 	6     	3.37568	0.756833	2.67377	4.56136


true minimum 2.76357717108


INFO:__main__:89 	6     	3.51921	0.73003 	2.67377	4.64387


true minimum 2.67376895409


INFO:__main__:90 	5     	3.74916	0.72867 	2.67377	4.53191


true minimum 2.67376895409


INFO:__main__:91 	5     	4.2997 	0.409688	3.89017	5.09722


true minimum 2.67376895409


INFO:__main__:92 	6     	3.58176	0.9467  	2.67377	5.07645


true minimum 2.67376895409


INFO:__main__:93 	6     	3.70092	0.558363	2.57566	4.19994


true minimum 2.57566379294


INFO:__main__:94 	6     	4.73009	0.381539	4.02488	5.07645


true minimum 2.57566379294


INFO:__main__:95 	6     	3.78595	0.434481	2.89454	4.22599


true minimum 2.57566379294


INFO:__main__:96 	6     	4.22617	0.18045 	3.91835	4.51524


true minimum 2.57566379294


INFO:__main__:97 	4     	3.88434	0.821507	2.57566	4.93564


true minimum 2.57566379294


INFO:__main__:98 	4     	4.45176	0.807055	3.08223	5.58645


true minimum 2.57566379294


INFO:__main__:99 	5     	3.13081	0.467016	2.57566	3.92952


true minimum 2.57566379294


INFO:__main__:100	3     	3.23386	0.278715	2.8643 	3.67439


true minimum 2.57566379294


INFO:__main__:101	6     	3.75427	0.643775	2.86317	4.7178 


true minimum 2.57566379294


INFO:__main__:102	5     	3.98314	0.547056	3.26943	4.59543


true minimum 2.57566379294


INFO:__main__:103	6     	3.84598	0.476044	3.12788	4.50232


true minimum 2.57566379294


INFO:__main__:104	6     	3.70835	0.554704	2.84278	4.52018


true minimum 2.57566379294


INFO:__main__:105	5     	3.56966	0.699784	2.57566	4.75566


true minimum 2.57566379294


INFO:__main__:106	6     	3.91155	0.531345	3.2797 	4.63286


true minimum 2.57566379294


INFO:__main__:107	6     	4.21167	0.359962	3.82872	4.69631


true minimum 2.57566379294


INFO:__main__:108	5     	3.49911	0.624585	2.51113	4.28619


true minimum 2.57566379294


INFO:__main__:109	5     	3.93379	0.504683	3.04739	4.50143


true minimum 2.57566379294


INFO:__main__:110	6     	3.80815	0.859721	2.5838 	4.84365


true minimum 2.57566379294


INFO:__main__:111	3     	4.44002	0.694086	2.95128	5.0606 


true minimum 2.57566379294


INFO:__main__:112	6     	3.51671	0.838114	2.71379	4.74616


true minimum 2.57566379294


INFO:__main__:113	6     	3.35559	0.696034	2.68253	4.65919


true minimum 2.57566379294


INFO:__main__:114	5     	4.06026	0.35726 	3.51036	4.48074


true minimum 2.57566379294


INFO:__main__:115	6     	3.7781 	0.624878	2.69079	4.53203


true minimum 2.57566379294


INFO:__main__:116	6     	4.52828	0.350257	4.02333	4.97471


true minimum 2.57566379294


INFO:__main__:117	6     	3.79893	0.733294	2.57566	4.73152


true minimum 2.57566379294


INFO:__main__:118	6     	4.12033	0.43679 	3.5553 	4.78582


true minimum 2.57566379294


INFO:__main__:119	6     	4.30814	0.606473	3.27424	5.11783


true minimum 2.57566379294


INFO:__main__:120	6     	4.65842	0.417121	4.20234	5.26784


true minimum 2.57566379294


INFO:__main__:121	4     	3.82854	1.2548  	2.57566	5.30091


true minimum 2.57566379294


INFO:__main__:122	6     	4.02754	0.928535	2.60214	5.08902


true minimum 2.57566379294


INFO:__main__:123	5     	3.73497	0.660277	2.92228	4.96072


true minimum 2.57566379294


INFO:__main__:124	6     	3.19457	0.771896	2.57566	4.68117


true minimum 2.57566379294


INFO:__main__:125	5     	3.10875	0.368865	2.75811	3.88946


true minimum 2.57566379294


INFO:__main__:126	6     	3.48925	0.482242	2.52307	3.99873


true minimum 2.52307142918


INFO:__main__:127	5     	4.19127	0.339083	3.74789	4.64646


true minimum 2.52307142918


INFO:__main__:128	6     	4.40673	0.395357	3.65162	4.75879


true minimum 2.52307142918


INFO:__main__:129	6     	4.16878	0.66866 	3.24772	5.17818


true minimum 2.52307142918


INFO:__main__:130	5     	4.27789	0.459872	3.78188	4.96411


true minimum 2.52307142918


INFO:__main__:131	6     	4.56138	0.395438	3.7653 	4.96354


true minimum 2.52307142918


INFO:__main__:132	5     	4.43038	0.394408	3.81467	4.85663


true minimum 2.52307142918


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:133	6     	4.14342	0.644854	2.82012	4.6975 


true minimum 2.52307142918


INFO:__main__:134	6     	4.25484	0.387427	3.90716	4.97103


true minimum 2.52307142918


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:135	6     	3.48361	0.807436	2.52307	4.6866 


true minimum 2.52307142918


INFO:__main__:136	4     	3.98643	0.897139	2.82012	5.02655


true minimum 2.52307142918


INFO:__main__:137	3     	3.68158	0.525409	2.82012	4.36394


true minimum 2.52307142918


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:138	6     	4.05625	0.440638	3.3411 	4.51551


true minimum 2.52307142918


/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in exp
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
/home/jovyan/neuronunit/neuronunit/tests/passive.py:85: RuntimeWarning: overflow encountered in multiply
  vm_fit[offset:,0] = a * np.exp(-t[offset:]/b) + c
INFO:__main__:139	6     	3.91535	0.503572	3.03509	4.36077


true minimum 2.52307142918


INFO:__main__:140	5     	3.87695	0.58941 	2.96266	4.5847 


true minimum 2.52307142918


INFO:__main__:141	6     	3.69296	0.565951	3.03827	4.46658


true minimum 2.52307142918


INFO:__main__:142	6     	3.29519	0.437201	2.71273	3.98314


true minimum 2.52307142918


INFO:__main__:143	6     	4.07268	0.334234	3.35507	4.36123


true minimum 2.52307142918


INFO:__main__:144	5     	3.96881	0.702166	2.68926	4.74622


true minimum 2.52307142918


INFO:__main__:145	6     	4.0623 	0.283412	3.50879	4.41424


true minimum 2.52307142918


INFO:__main__:146	6     	4.21897	0.474539	3.18645	4.58636


true minimum 2.52307142918


INFO:__main__:147	5     	4.28308	0.376596	3.66355	4.82235


true minimum 2.52307142918


INFO:__main__:148	5     	4.11082	0.0954588	4.06037	4.32177


true minimum 2.52307142918


INFO:__main__:149	6     	3.87063	0.657622 	2.80581	4.47962


true minimum 2.52307142918


INFO:__main__:150	6     	3.95304	0.490059 	3.23952	4.40974


true minimum 2.52307142918


The optimizer needs to find a parameter set much better than 4.66 in order to be worthwhile, for the case of the Neo Cortical Pyramidal neuron. In only 10 generations the optimizer finds a parameter set 3.24. In 23 generations it finds a model with score 2.99.

In [77]:
for k,v in test_opt.items():
    #list_best = []
    best_model = v['pf'][0]
    second_best_model = v['pf'][1]
    third_best_model = v['pf'][2]
    list_best = [best_model,second_best_model,third_best_model]
summed = best_model.dtc.get_ss()
print('its {2} that {0} is significantly less than {1}'.format(summed,5.47,summed<5.47))
print(key)

its True that 2.523071429176899 is significantly less than 5.47
Neocortex pyramidal cell layer 5-6


In [44]:
best_model.dtc.attrs

{'C': 80.32327976522879,
 'a': 0.017035973642268745,
 'b': -1.995763503718458,
 'c': -56.110284353607987,
 'd': 39.870524431534278,
 'k': 0.52366419898375982,
 'vPeak': 3.4684930072840432,
 'vr': -67.290036991963049}

In [45]:
best_model.dtc.scores


{'CapacitanceTest': 0.14163892773420128,
 'InjectedCurrentAPAmplitudeTest': 0.9999999997243523,
 'InjectedCurrentAPThresholdTest': 0.3846682140841715,
 'InjectedCurrentAPWidthTest': 0.10023627982283978,
 'InputResistanceTest': 0.08782389566447724,
 'RestingPotentialTest': 0.02152619567192071,
 'RheobaseTestP': 0.7321282418074668,
 'TimeConstantTest': 0.05504967466746935}

In [50]:
best_model.dtc.score

{'CapacitanceTest': {'value': 0.8583610722657987},
 'InjectedCurrentAPAmplitudeTest': {'value': 2.7564772686616834e-10},
 'InjectedCurrentAPThresholdTest': {'value': 0.6153317859158285},
 'InjectedCurrentAPWidthTest': {'value': 0.8997637201771602},
 'InputResistanceTest': {'value': 0.9121761043355228},
 'RestingPotentialTest': {'value': 0.9784738043280793},
 'RheobaseTestP': {'value': 0.26787175819253317},
 'TimeConstantTest': {'value': 0.9449503253325306}}

In [49]:
for row in test_frame['Neocortex pyramidal cell layer 5-6']:
    print(row.observation)


{'mean': array(213.849583333333) * pA, 'n': 32, 'std': array(170.452454715608) * pA}
{'mean': array(120.672073643411) * Mohm, 'n': 86, 'std': array(77.6331608333564) * Mohm}
{'mean': array(15.7342424242424) * ms, 'n': 33, 'std': array(7.31162636832495) * ms}
{'mean': array(150.584166666667) * pF, 'n': 12, 'std': array(139.683884626343) * pF}
{'mean': array(-68.2481434599156) * mV, 'n': 79, 'std': array(6.53234788156637) * mV}
{'mean': array(1.20769387755102) * ms, 'n': 49, 'std': array(0.534345918375033) * ms}
{'mean': array(80.4351020408164) * mV, 'n': 49, 'std': array(12.7488030357545) * mV}
{'mean': array(-42.7357232704403) * mV, 'n': 53, 'std': array(8.04073233409085) * mV}


In [51]:
key = str('Neocortex pyramidal cell layer 5-6')



'''
dtc = dtc_to_rheo(dtc)
dtc = format_test(dtc)
if dtc.rheobase is not None:
    if dtc.rheobase!=-1.0:
        dtc = nunit_evaluation(dtc)
'''        
    


In [76]:
#df

In [83]:
from neuronunit.optimization.optimization_management import bridge_judge
dfs = []
for model in list_best:
    df = pd.DataFrame(index=list(dtc.tests),columns=['observation','prediction','disagreement'])#,columns=list(reduced_cells.keys()))

    dtc = DataTC()
    dtc.tests = use_test
    dtc.attrs = model.dtc.attrs
    dtc.backend = 'RAW'
    dtc.cell_name = 'vanilla'
    #for key, use_test in test_frame.items():
    dtc.tests = test_frame[key]
    dtc = dtc_to_rheo(dtc)
    dtc = format_test(dtc)
    for k,t in enumerate(dtc.tests):
        t.params = dtc.vtest[k]
        score,pred = bridge_judge((t,dtc))
        #print(score,pred)       
        df.iloc[k]['observation'] = t.observation['mean'] 
        try:
            agreement = np.abs(t.observation['mean'] - pred['value'])
            df.iloc[k]['prediction'] = pred['value']
            df.iloc[k]['disagreement'] = agreement

        except:
            agreement = np.abs(t.observation['mean'] - pred['mean'])
            df.iloc[k]['prediction'] = pred['mean']
            df.iloc[k]['disagreement'] = agreement
    dfs.append(df)
result = pd.concat(dfs)
result         

observation  \
RheobaseTestP                     213.849583333333 pA   
InputResistanceTest             120.672073643411 Mohm   
TimeConstantTest                  15.7342424242424 ms   
CapacitanceTest                   150.584166666667 pF   
RestingPotentialTest             -68.2481434599156 mV   
InjectedCurrentAPWidthTest        1.20769387755102 ms   
InjectedCurrentAPAmplitudeTest    80.4351020408164 mV   
InjectedCurrentAPThresholdTest   -42.7357232704403 mV   
RheobaseTestP                     213.849583333333 pA   
InputResistanceTest             120.672073643411 Mohm   
TimeConstantTest                  15.7342424242424 ms   
CapacitanceTest                   150.584166666667 pF   
RestingPotentialTest             -68.2481434599156 mV   
InjectedCurrentAPWidthTest        1.20769387755102 ms   
InjectedCurrentAPAmplitudeTest    80.4351020408164 mV   
InjectedCurrentAPThresholdTest   -42.7357232704403 mV   
RheobaseTestP                     213.849583333333 pA   
InputResistanceTest             120.672073643411 Mohm   
TimeConstantTest                  15.7342424242424 ms   
CapacitanceTest                   150.584166666667 pF   
RestingPotentialTest             -68.2481434599156 mV   
InjectedCurrentAPWidthTest        1.20769387755102 ms   
InjectedCurrentAPAmplitudeTest    80.4351020408164 mV   
InjectedCurrentAPThresholdTest   -42.7357232704403 mV   

                                                                prediction  \
RheobaseTestP                                                24.9921875 pA   
InputResistanceTest                 129234556.37052076 kg*m**2/(s**3*A**2)   
TimeConstantTest                                    0.016239105584981928 s   
CapacitanceTest                 1.2565606321596948e-10 s**4*A**2/(kg*m**2)   
RestingPotentialTest                                -0.06842440161605753 V   
InjectedCurrentAPWidthTest                                      0.001275 s   
InjectedCurrentAPAmplitudeTest                    -3.394359574052097e-05 V   
InjectedCurrentAPThresholdTest                      -0.03869543745051891 V   
RheobaseTestP                                              25.154296875 pA   
InputResistanceTest                 127521728.72962098 kg*m**2/(s**3*A**2)   
TimeConstantTest                                    0.015195661560195011 s   
CapacitanceTest                 1.1916135165022537e-10 s**4*A**2/(kg*m**2)   
RestingPotentialTest                                  -0.068399020199018 V   
InjectedCurrentAPWidthTest                                      0.001275 s   
InjectedCurrentAPAmplitudeTest                    -0.0003786137860437716 V   
InjectedCurrentAPThresholdTest                      -0.03830641512782343 V   
RheobaseTestP                                                24.9921875 pA   
InputResistanceTest                 129375643.30537802 kg*m**2/(s**3*A**2)   
TimeConstantTest                                     0.01617715819692096 s   
CapacitanceTest                    1.2504021455365e-10 s**4*A**2/(kg*m**2)   
RestingPotentialTest                                -0.06839796225443817 V   
InjectedCurrentAPWidthTest                         0.0012000000000000001 s   
InjectedCurrentAPAmplitudeTest                     -0.001715732173261715 V   
InjectedCurrentAPThresholdTest                      -0.03680516308404327 V   

                                           disagreement  
RheobaseTestP                       188.857395833333 pA  
InputResistanceTest              8.562482727109753 Mohm  
TimeConstantTest                  0.5048631607395286 ms  
CapacitanceTest                    24.92810345069752 pF  
RestingPotentialTest             0.17625815614194096 mV  
InjectedCurrentAPWidthTest       0.06730612244898015 ms  
InjectedCurrentAPAmplitudeTest     80.46904563655693 mV  
InjectedCurrentAPThresholdTest     4.040285819921387 mV  
RheobaseTestP                       188.695286458333 pA  
InputResistanceTest              6.849655086209978 Mohm  
TimeConstantTest           

In [ ]:
df
from neuronunit.models.reduced import ReducedModel
from neuronunit.optimization.data_transport_container import DataTC
from neuronunit.optimization.model_parameters import model_params, path_params, transcribe_units
from neuronunit.optimization.optimization_management import mint_generic_model
LEMS_MODEL_PATH = path_params['model_path']


model = ReducedModel(LEMS_MODEL_PATH,name = str('vanilla'),backend = ('RAW'))
dtc = DataTC()
dtc.attrs = model.attrs